### ISSUE TITLE
Phase 11.3 — ValueLearner: reward model fine-tuning from human feedback signals

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#343]
## Phase 11.3 — `ValueLearner`: Reward Model Fine-Tuning from Human Feedback Signals

**Phase**: 11 — Safety & Alignment
**Depends on**: #340 (AlignmentMonitor), #337 (SafetyFilter)
**Milestone**: Phase 11 complete when all 5 sub-phases are spec'd

---

## Motivation

`AlignmentMonitor` (#340) provides *continuous* alignment health scoring against pre-defined constitutional dimensions. `ValueLearner` closes the human-in-the-loop: it captures explicit feedback signals (approve / reject / adjust) from human operators and federation peers, fits a lightweight reward model from those signals, and propagates updated value weights back into the `SafetyFilter` ruleset and `AlignmentMonitor` dimension thresholds.

Without `ValueLearner`, alignment rules are static — they can only be updated by manual config changes. With `ValueLearner`, the system learns *which behaviors humans actually prefer* rather than relying solely on upfront rule authorship.

---

### `FeedbackSignal` enum (6 signal types)

```python
class FeedbackSignal(Enum):
    APPROVE          = "approve"          # human explicitly approves action
    REJECT           = "reject"           # human explicitly rejects action
    ADJUST_WEIGHT    = "adjust_weight"    # direct scalar weight nudge
    COMPARATIVE      = "comparative"      # prefer A over B (pair ranking)
    IMPLICIT_POSITIVE = "implicit_pos"   # inferred from positive outcome
    IMPLICIT_NEGATIVE = "implicit_neg"   # inferred from negative outcome
```

### Frozen dataclasses

```python
@dataclass(frozen=True)
class FeedbackEntry:
    entry_id: str                    # UUID
    goal_id: str
    dimension: AlignmentDimension    # which alignment axis
    signal: FeedbackSignal
    value: float                     # scalar in [-1.0, 1.0]
    annotator_id: str                # human or peer DID
    timestamp: float                 # epoch seconds
    context_hash: str                # sha256 of (goal_id + task_ids)

@dataclass(frozen=True)
class RewardModelWeights:
    weights: tuple[float, ...]       # len == len(AlignmentDimension)
    bias: float
    version: int
    trained_at: float
    sample_count: int

@dataclass(frozen=True)
class ValueLearnerConfig:
    learning_rate: float = 0.01
    regularisation_lambda: float = 1e-4
    min_feedback_to_train: int = 10  # minimum samples before first model update
    max_history: int = 10_000        # FIFO eviction of oldest feedback
    update_interval_s: float = 60.0  # background model update period
    comparative_margin: float = 0.1  # margin for comparative ranking loss

@dataclass(frozen=True)
class ValueLearnerSnapshot:
    total_feedback: int
    model_version: int
    weights: tuple[float, ...]
    bias: float
    last_trained_at: float | None
    pending_feedback: int
```

---

## `ValueLearner` Protocol

```python
class ValueLearner(Protocol):
    async def record_feedback(self, entry: FeedbackEntry) -> None: ...
    async def train(self) -> RewardModelWeights: ...
    async def score(self, dimension: AlignmentDimension, context_hash: str) -> float: ...
    def current_weights(self) -> RewardModelWeights: ...
    def snapshot(self) -> ValueLearnerSnapshot: ...
```

---

### Fields

```python
class InMemoryValueLearner:
    _config: ValueLearnerConfig
    _history: collections.deque[FeedbackEntry]   # maxlen=max_history
    _weights: list[float]                        # len == 5 (one per AlignmentDimension)
    _bias: float
    _version: int
    _last_trained_at: float | None
    _pending: list[FeedbackEntry]                # since last train() call
    _lock: asyncio.Lock
    _bg_task: asyncio.Task | None
```

### Key methods

**`record_feedback(entry)`**
- Acquire `_lock`
- Append to `_history` (deque auto-evicts oldest)
- Append to `_pending`
- Release lock

**`_feature_vector(entry) -> list[float]`**
- One-hot encode `entry.dimension` (5 dims)
- Append `entry.value` as 6th feature
- Return 6-element list

**`train() -> RewardModelWeights`**
- Skip if `len(_history) < config.min_feedback_to_train`
- Build training set from `_history`:
  - APPROVE/REJECT/ADJUST_WEIGHT/IMPLICIT_*: supervised MSE target = `entry.value`
  - COMPARATIVE pairs: ranking loss — ensure `score(A) > score(B) + margin`
- **Gradient descent** (stochastic, one epoch over `_pending`):
  ```python
  for entry in self._pending:
      x = self._feature_vector(entry)
      pred = dot(self._weights, x[:5]) + self._bias
      target = _signal_to_target(entry.signal, entry.value)
      err = pred - target
      for i in range(5):
          self._weights[i] -= lr * (err * x[i] + λ * self._weights[i])
      self._bias -= lr * err
  ```
- Clip weights to `[-2.0, 2.0]` (stability guard)
- Increment `_version`, set `_last_trained_at`, clear `_pending`
- Return frozen `RewardModelWeights`

**`score(dimension, context_hash) -> float`**
- Compute `dot(weights, one_hot(dimension)) + bias`
- Clamp to `[0.0, 1.0]`

**`_bg_update_loop()`**
- Every `update_interval_s`: call `await self.train()`
- Log `model_version` and `pending_feedback` count

**`start() / stop()`**
- Launch / cancel `_bg_update_loop` task

---

## Factory

```python
def build_value_learner(config: ValueLearnerConfig | None = None) -> ValueLearner:
    return InMemoryValueLearner(config or ValueLearnerConfig())
```

---

## `CognitiveCycle` Integration

```python
class CognitiveCycle:
    _value_learner: ValueLearner

    async def _handle_goal_outcome(self, goal: Goal, outcome: GoalStatus) -> None:
        signal = (
            FeedbackSignal.IMPLICIT_POSITIVE if outcome == GoalStatus.COMPLETED
            else FeedbackSignal.IMPLICIT_NEGATIVE
        )
        for dim in AlignmentDimension:
            entry = FeedbackEntry(
                entry_id=str(uuid4()),
                goal_id=goal.goal_id,
                dimension=dim,
                signal=signal,
                value=1.0 if signal == FeedbackSignal.IMPLICIT_POSITIVE else -0.5,
                annotator_id="cognitive_cycle",
                timestamp=time.time(),
                context_hash=hashlib.sha256(goal.goal_id.encode()).hexdigest()[:16],
            )
            await self._value_learner.record_feedback(entry)
```

---

## Prometheus Metrics

| Metric | Type | Description |
|--------|------|-------------|
| `value_learner_feedback_total` | Counter | Total feedback entries recorded, by `signal` label |
| `value_learner_model_version` | Gauge | Current reward model version number |
| `value_learner_train_duration_seconds` | Histogram | Time spent in `train()` call |
| `value_learner_pending_feedback` | Gauge | Feedback entries since last train |
| `value_learner_weight_norm` | Gauge | L2 norm of current weight vector |

### PromQL examples

```promql
# Feedback rate by signal type
rate(value_learner_feedback_total[5m])

# Model freshness (seconds since last train)
time() - value_learner_last_trained_timestamp

# Weight drift (L2 norm increasing → model diverging)
delta(value_learner_weight_norm[1h])
```

---

## mypy Compliance

| Symbol | Notes |
|--------|-------|
| `FeedbackSignal` | `StrEnum` |
| `FeedbackEntry` | `frozen=True`, all fields typed |
| `RewardModelWeights` | `tuple[float, ...]` immutable weights |
| `ValueLearnerConfig` | `frozen=True`, all defaults typed |
| `ValueLearner` | `Protocol` with `runtime_checkable` |
| `InMemoryValueLearner` | `asyncio.Lock`, `deque[FeedbackEntry]` |
| `build_value_learner` | returns `ValueLearner` (structural subtype) |

---

## Test Targets (12)

1. `test_record_feedback_appends_to_history`
2. `test_history_evicts_oldest_at_maxlen`
3. `test_train_skips_below_min_feedback`
4. `test_train_returns_frozen_weights`
5. `test_train_increments_version`
6. `test_score_clamps_to_unit_interval`
7. `test_approve_signal_increases_dimension_weight`
8. `test_reject_signal_decreases_dimension_weight`
9. `test_comparative_signal_rank_order`
10. `test_bg_update_loop_calls_train`
11. `test_snapshot_reflects_state`
12. `test_weight_clip_prevents_divergence`

---

## Implementation Order (14 steps)

1. Define `FeedbackSignal` enum
2. Define `FeedbackEntry`, `RewardModelWeights`, `ValueLearnerConfig`, `ValueLearnerSnapshot` dataclasses
3. Define `ValueLearner` Protocol
4. Stub `InMemoryValueLearner.__init__` with fields
5. Implement `record_feedback()` with deque + lock
6. Implement `_feature_vector()` one-hot encoder
7. Implement `_signal_to_target()` helper
8. Implement `train()` gradient descent pass + comparative ranking loss
9. Implement `score()` linear scorer with clamp
10. Implement `_bg_update_loop()` asyncio task
11. Implement `start()` / `stop()` lifecycle
12. Implement `snapshot()`
13. Implement `build_value_learner()` factory
14. Pre-register Prometheus metrics at import time

---

## Phase 11 sub-phase roadmap

| Sub-phase | Component | Status |
|-----------|-----------|--------|
| 11.1 | SafetyFilter — constitutional ruleset gating | ✅ Spec filed (#337) |
| 11.2 | AlignmentMonitor — continuous drift detection | ✅ Spec filed (#340) |
| 11.3 | ValueLearner — reward model from human feedback | 🟡 This issue |
| 11.4 | InterpretabilityProbe — activation patching & steering | 📋 Planned |
| 11.5 | AlignmentDashboard — unified safety/alignment UI | 📋 Planned |

**Comment by web3guru888** [comment]
## Implementation Notes

### `_feature_vector()` — One-Hot Encoding

```python
_DIM_INDEX = {d: i for i, d in enumerate(AlignmentDimension)}

def _feature_vector(self, entry: FeedbackEntry) -> list[float]:
    vec = [0.0] * 5
    vec[_DIM_INDEX[entry.dimension]] = 1.0
    vec.append(entry.value)   # 6th element: scalar signal magnitude
    return vec
```

### `_signal_to_target()` — Map Signal to Regression Target

```python
_SIGNAL_TARGETS: dict[FeedbackSignal, float] = {
    FeedbackSignal.APPROVE:            1.0,
    FeedbackSignal.REJECT:            -1.0,
    FeedbackSignal.IMPLICIT_POSITIVE:  0.7,
    FeedbackSignal.IMPLICIT_NEGATIVE: -0.5,
    FeedbackSignal.ADJUST_WEIGHT:      None,   # use entry.value directly
    FeedbackSignal.COMPARATIVE:        None,   # handled separately
}

def _signal_to_target(signal: FeedbackSignal, value: float) -> float:
    t = _SIGNAL_TARGETS.get(signal)
    return t if t is not None else value
```

### `train()` — Gradient Descent with Ranking Loss

```python
async def train(self) -> RewardModelWeights:
    async with self._lock:
        if len(self._history) < self._config.min_feedback_to_train:
            return self.current_weights()

        lr = self._config.learning_rate
        lam = self._config.regularisation_lambda
        batch = list(self._pending)

        for entry in batch:
            if entry.signal == FeedbackSignal.COMPARATIVE:
                continue  # handled in ranking pass below
            x = self._feature_vector(entry)
            pred = sum(w * f for w, f in zip(self._weights, x[:5])) + self._bias
            target = _signal_to_target(entry.signal, entry.value)
            err = pred - target
            for i in range(5):
                self._weights[i] -= lr * (err * x[i] + lam * self._weights[i])
            self._bias -= lr * err

        # Ranking loss pass for COMPARATIVE signals
        comparatives = [e for e in batch if e.signal == FeedbackSignal.COMPARATIVE]
        # (Pairs encoded as sequential entries; odd index = preferred, even = rejected)
        for i in range(0, len(comparatives) - 1, 2):
            pref, rej = comparatives[i], comparatives[i + 1]
            s_pref = sum(w * f for w, f in zip(
                self._weights, self._feature_vector(pref)[:5])) + self._bias
            s_rej  = sum(w * f for w, f in zip(
                self._weights, self._feature_vector(rej)[:5]))  + self._bias
            margin_viol = max(0.0, s_rej - s_pref + self._config.comparative_margin)
            if margin_viol > 0:
                for i in range(5):
                    fp = self._feature_vector(pref)
                    fr = self._feature_vector(rej)
                    self._weights[i] -= lr * (fr[i] - fp[i])

        # Clip weights for stability
        self._weights = [max(-2.0, min(2.0, w)) for w in self._weights]
        self._version += 1
        self._last_trained_at = time.time()
        self._pending.clear()

        return RewardModelWeights(
            weights=tuple(self._weights),
            bias=self._bias,
            version=self._version,
            trained_at=self._last_trained_at,
            sample_count=len(self._history),
        )
```

### mypy Table

| Symbol | mypy flag |
|--------|-----------|
| `FeedbackEntry.context_hash` | `str` (not `bytes`) |
| `RewardModelWeights.weights` | `tuple[float, ...]` (not `list`) |
| `deque[FeedbackEntry]` | typed generic |
| `asyncio.Lock` | no `Optional` needed — always constructed |
| `_bg_task: asyncio.Task | None` | union with `None` sentinel |

### Test Isolation Skeleton

```python
@pytest.fixture
def learner() -> InMemoryValueLearner:
    return InMemoryValueLearner(
        ValueLearnerConfig(min_feedback_to_train=2, max_history=5)
    )

@pytest.mark.asyncio
async def test_approve_signal_increases_dimension_weight(learner):
    dim = AlignmentDimension.CONSTITUTIONAL
    before = learner._weights[_DIM_INDEX[dim]]
    for _ in range(3):
        await learner.record_feedback(FeedbackEntry(
            entry_id=str(uuid4()), goal_id="g1", dimension=dim,
            signal=FeedbackSignal.APPROVE, value=1.0,
            annotator_id="human", timestamp=time.time(), context_hash="abc",
        ))
    await learner.train()
    assert learner._weights[_DIM_INDEX[dim]] > before
```

### 14-Step Implementation Order

1. `FeedbackSignal` enum
2. Frozen dataclasses (`FeedbackEntry`, `RewardModelWeights`, `ValueLearnerConfig`, `ValueLearnerSnapshot`)
3. `ValueLearner` Protocol
4. `_DIM_INDEX` mapping constant
5. `InMemoryValueLearner.__init__` with all fields
6. `record_feedback()` + `asyncio.Lock`
7. `_feature_vector()` + `_signal_to_target()`
8. `train()` MSE pass
9. `train()` comparative ranking pass + weight clipping
10. `score()` linear + clamp
11. `_bg_update_loop()` asyncio task
12. `start()` / `stop()` lifecycle
13. `snapshot()`
14. `build_value_learner()` factory + Prometheus pre-registration

**Comment by web3guru888** [comment]
## Phase 11.4 filed — InterpretabilityProbe

Issue #346 tracks **Phase 11.4 — InterpretabilityProbe**: feature attribution and explanation generation for alignment decisions.

### ValueLearner → InterpretabilityProbe integration

`InterpretabilityProbe` can explain `ValueLearner` weight updates via `ExplanationTarget.REWARD_WEIGHT`:
- Feature vector: `FeedbackEntry` fields (signal one-hot, dimension one-hot, value scalar, annotator_id hash)
- Decision value: weight delta magnitude after a training step
- Attribution output: which feedback signals most influenced the weight update

This provides **audit trails** for reward model changes — operators can see exactly which human feedback drove a weight shift.

### Phase 11 sub-phase tracker

| Sub-phase | Component | Issue | Status |
|-----------|-----------|-------|--------|
| 11.1 | SafetyFilter | #337 | ✅ spec'd |
| 11.2 | AlignmentMonitor | #340 | ✅ spec'd |
| 11.3 | ValueLearner | #343 | ✅ spec'd |
| 11.4 | InterpretabilityProbe | #346 | 🟡 spec'd |
| 11.5 | AlignmentDashboard | planned | ⏳ next |
